go/vt/mysqlctl/xtrabackupengine.go

```diff
@@ -203,11 +210,37 @@ func (be *XtrabackupEngine) backupFiles(ctx context.Context, cnf *Mycnf, logger
 		flagsToExec = append(flagsToExec, strings.Fields(*xtrabackupBackupFlags)...)
 	}
 
-	destFiles, err := addStripeFiles(ctx, bh, backupFileName, numStripes, logger)
+	// Create a cancellable Context for calls to bh.AddFile().
+	// This allows us to decide later if we need to cancel an attempt to Close()
+	// the file returned from AddFile(), since Close() itself does not accept a
+	// Context value. We can't use a context.WithTimeout() here because that
+	// would impose a timeout that starts counting right now, so it would
+	// include the time spent uploading the file content. We only want to impose
+	// a timeout on the final Close() step.
+	addFilesCtx, cancelAddFiles := context.WithCancel(ctx)
+	defer cancelAddFiles()
+	destFiles, err := addStripeFiles(addFilesCtx, bh, backupFileName, numStripes, logger)
 	if err != nil {
 		return replicationPosition, vterrors.Wrapf(err, "cannot create backup file %v", backupFileName)
 	}
 	defer func() {
+		// Impose a timeout on the process of closing files.
+		go func() {
+			timer := time.NewTimer(closeTimeout)
+
+			select {
+			case <-addFilesCtx.Done():
+				timer.Stop()
+				return
+			case <-timer.C:
+				logger.Errorf("Timed out waiting for Close() on backup file to complete")
+				// Cancelling the Context that was originally passed to bh.AddFile()
+				// should hopefully cause Close() calls on the file that AddFile()
+				// returned to abort.
+				cancelAddFiles()
+			}
+		}()
```

Comment:
I was about to merge, but... I'm not sure this solves the problem as described. The above `go func` will cancel the context and exit on the time out, but the outer `defer func` can still hang because `closeFile` doesn't care about `cancelAddFiles`.

Did you mean to do it the other way round, where the `closeFile` runs in a goroutine but the `defer func` should exit on timeout without waiting for `closeFile` to finish?

Reply:
If we fail to make `Close()` return, then abandoning its goroutine will leak it, which could have unpredictable side effects down the line. It seems safer to me to hang if `Close()` hangs. We could also force a hard crash, but I'd rather give this a try first and add the crash only if we really need it. WDYT?

Reply:
I added a comment to document the above caveat.